### PR TITLE
WIP: Pick place capability (code review)

### PR DESCRIPTION
**Pull request by felixvd**
This PR isn't for merging, it's just for discussion and commenting. There might be a better way, but I don't know about it.

### CODE REVIEW

File: core/include/moveit/task_constructor/stages/pick.h [hunk 1]
```diff
@@ -53,73 +61,166 @@ namespace stages {
  *
  * Picking consist of the following sub stages:
  * - linearly approaching the object along an approach direction/twist
- * - "grasp" end effector posture
+ * - GraspProviderPlugin stage wrapped in an IK computation
```
Comment (felixvd) [inline core/include/moveit/task_constructor/stages/pick.h]: I don't understand the "wrapped in an IK computation" part

File: core/include/moveit/task_constructor/stages/pick.h [hunk 2]
```diff
@@ -53,73 +61,166 @@ namespace stages {
...
  * - detach the object
  * - linearly retract end effector
- *
- * The end effector postures corresponding to pre-grasp and grasp as well as
- * the end effector's Cartesian pose needs to be provided by an external grasp stage.
  */
 class PickPlaceBase : public SerialContainer
 {
+
+	bool is_pick_;
```
Comment (felixvd) [inline core/include/moveit/task_constructor/stages/pick.h]: This needs a comment or explanation, it's very hard to follow how this stage is meant to be read.

File: core/include/moveit/task_constructor/stages/pick.h [hunk 3]
```diff
@@ -53,73 +61,166 @@ namespace stages {
...
+
 	solvers::CartesianPathPtr cartesian_solver_;
-	Stage* grasp_stage_ = nullptr;
-	Stage* approach_stage_ = nullptr;
-	Stage* lift_stage_ = nullptr;
+	solvers::PipelinePlannerPtr sampling_planner_;
+
+	Stage* move_there_stage_ = nullptr;
```
Comment (felixvd) [inline core/include/moveit/task_constructor/stages/pick.h]: Why not keep the name `approach_stage_`?

File: core/include/moveit/task_constructor/stages/pick.h [hunk 4]
```diff
@@ -53,73 +61,166 @@ namespace stages {
...
 
 	void init(const moveit::core::RobotModelConstPtr& robot_model) override;
 
+	/// -------------------------
+	/// setters of own properties
+
+	/// eef
 	void setEndEffector(const std::string& eef) { properties().set<std::string>("eef", eef); }
-	void setObject(const std::string& object) { properties().set<std::string>("object", object); }
+	void setEndEffectorOpenClose(const std::string& open_pose_name, const std::string& close_pose_name){
```
Comment (felixvd) [inline core/include/moveit/task_constructor/stages/pick.h]: Would prefer the name `setEndEffectorOpenCloseProperties` or `setEndEffectorOpenClosePoses`

File: core/include/moveit/task_constructor/stages/pick.h [hunk 5]
```diff
@@ -53,73 +61,166 @@ namespace stages {
...
+		pose = p;
+		setIKFrame(pose, link);
+	}
+	void setIKFrame(const std::string& link) { setIKFrame(Eigen::Isometry3d::Identity(), link); }
+
+	/// -------------------------
+	/// setters of substage properties
+
+	/// approach / place
+	void setApproachPlace(const geometry_msgs::TwistStamped& motion, double min_distance, double max_distance);
```
Comment (felixvd) [inline core/include/moveit/task_constructor/stages/pick.h]: Would name this `setApproachDirection`. More verbosity will help.

File: core/include/moveit/task_constructor/stages/pick.h [hunk 6]
```diff
@@ -53,73 +61,166 @@ namespace stages {
...
+	void setApproachPlace(const geometry_msgs::Vector3Stamped& direction, double min_distance, double max_distance);
+	void setApproachPlace(const std::map<std::string, double>& joints);
+
+	/// IK computation
+	void setMaxIKSolutions(const uint32_t& max_ik_solutions);
+	void setMinIKSolutionDistance(const double& min_ik_solution_distance);
+	void setIgnoreIKCollisions(const bool& ignore_ik_collisions);
+
+	/// lift / retract
+	void setLiftRetract(const geometry_msgs::TwistStamped& motion, double min_distance, double max_distance);
```
Comment (felixvd) [inline core/include/moveit/task_constructor/stages/pick.h]: `setRetreatDirection`

Comment (felixvd) [reply core/include/moveit/task_constructor/stages/pick.h]: I guess `setRetractDirection` works, too. I found `retreat` to be clearer, but it's a matter of opinion.

File: core/src/stages/pick.cpp [hunk 7]
```diff
@@ -1,63 +1,143 @@
...
-		pose.header.frame_id = boost::any_cast<std::string>(frame);
-		pose.pose.orientation.w = 1.0;
-		return pose;
-	};
+	provider_stage_plugin_name_ = provider_stage_plugin_name;
 
 	{
-		auto approach = std::make_unique<MoveRelative>(forward ? "approach object" : "retract", cartesian_solver_);
-		PropertyMap& p = approach->properties();
+		auto move_there = std::make_unique<MoveRelative>(is_pick_ ? "approach object" : "lower object", cartesian_solver_);
```
Comment (felixvd) [inline core/src/stages/pick.cpp]: `move_there` can be read in an imperative tone, so the intent is not clear upon reading. I'd even call this `approach_stage`, since you also assign it to `approach_stage_` a few lines later.

File: core/src/stages/pick.cpp [hunk 8]
```diff
@@ -1,63 +1,143 @@
...
+		p.property("ik_frame").configureInitFrom(Stage::PARENT, "ik_frame");
+		p.set("marker_ns", std::string(is_pick_ ? "approach" : "place"));
+		move_there_stage_ = move_there.get();
+		insert(std::move(move_there));
+	}
+
+	if (provider_stage_plugin_name_ == "")
+	{
+		if (is_pick_)
+			provider_stage_plugin_name_ = "moveit_task_constructor/GraspProviderDefault";
```
Comment (felixvd) [inline core/src/stages/pick.cpp]: `GraspProviderDefault` actually provides a *Grasp*, not just a pose, right?

File: core/src/stages/pick.cpp [hunk 9]
```diff
@@ -1,63 +1,143 @@
...
+		else
+			provider_stage_plugin_name_ = "moveit_task_constructor/PlaceProviderDefault";
```
Comment (felixvd) [inline core/src/stages/pick.cpp]: I'd call it `PlacePoseProviderDefault` if that hasn't already been decided somewhere else.

File: core/src/stages/pick.cpp [hunk 10]
```diff
@@ -1,63 +1,143 @@
...
+		insert(std::move(wrapper));
+
+	}
+	catch(pluginlib::PluginlibException& ex)
+	{
+		ROS_ERROR("The provider stage plugin failed to load. Error: %s", ex.what());
+	}
+
+	{
+		auto set_collision_object_hand = std::make_unique<stages::ModifyPlanningScene>(is_pick_ ? "allow collision (hand,object)" : "forbid collision (hand,object)");
```
Comment (felixvd) [inline core/src/stages/pick.cpp]: As above, I would add `_stage` to all of these temporary variables as well.

File: core/src/stages/pick.cpp [hunk 11]
```diff
@@ -1,63 +1,143 @@
...
+	if (is_pick_) {
+		auto set_collision_object_support = std::make_unique<stages::ModifyPlanningScene>("allow collision (object,support)");
+		allow_collision_object_support_stage_ = set_collision_object_support.get();
+		insert(std::move(set_collision_object_support));
+	}
 
 	{
-		auto lift = std::make_unique<MoveRelative>(forward ? "lift object" : "place object", cartesian_solver_);
-		PropertyMap& p = lift->properties();
+		auto move_back = std::make_unique<MoveRelative>(is_pick_ ? "lift object" : "retract", cartesian_solver_);
```
Comment (felixvd) [inline core/src/stages/pick.cpp]: `retreat_stage` or `retract_stage`

File: core/include/moveit/task_constructor/tasks/pick_place_task.h [hunk 12]
```diff
@@ -83,22 +85,27 @@ class PickPlaceTask
 
     // object + surface
     std::vector<std::string> support_surfaces_;
-    std::string object_reference_frame_;
-    std::string surface_link_;
+    // std::string object_reference_frame_;
```
Comment (felixvd) [inline core/include/moveit/task_constructor/tasks/pick_place_task.h]: Don't leave commented code. Just delete what's unnecessary.

File: core/src/tasks/pick_place_task.cpp [hunk 13]
```diff
@@ -131,254 +131,54 @@ void PickPlaceTask::init(const Parameters& parameters)
...
-   *                                                    *
-   *          Move to Place                             *
-   *                                                    *
-   *****************************************************/
-  {
-    auto stage = std::make_unique<stages::Connect>(
-        "move to place", stages::Connect::GroupPlannerVector{ { parameters.arm_group_name_, sampling_planner } });
-    stage->setTimeout(5.0);
-    stage->properties().configureInitFrom(Stage::PARENT);
+    auto stage = std::make_unique<stages::Pick>("Pick object", parameters.grasp_provider_plugin_name_);
```
Comment (felixvd) [inline core/src/tasks/pick_place_task.cpp]: This seems a lot more difficult to read now. I also don't know that this example code is the right place to explain how to use the capability (I assume that that's what you're trying to do). The MTC example code would still demonstrate the MTC internals and how to make new tasks. End users who just want to use the PickPlace capability with the `move_group_interface` would look at the MoveIt tutorial. I would update the `PickPlace` tutorial there instead.

File: capabilities/src/plan_pick_place_capability.cpp [hunk 14]
```diff
@@ -60,12 +60,31 @@ void PlanPickPlaceCapability::goalCallback(
...
+  parameters.hand_group_name_ = goal->hand_group_name;
+  parameters.eef_name_ = goal->eef_name;
+  parameters.hand_frame_ = goal->hand_frame;
+  parameters.object_name_ = goal->object_id;
+  parameters.support_surfaces_ = goal->support_surfaces;
+  parameters.grasp_provider_plugin_name_ = goal->grasp_provider_plugin_name;
+  tf::poseMsgToEigen(goal->grasp_frame_transform, parameters.grasp_frame_transform_);
+  parameters.hand_open_pose_ = "open";
+  parameters.hand_close_pose_ = "close";
+  parameters.approach_object_direction_ = goal->grasp.pre_grasp_approach.direction;
```
Comment (felixvd) [inline capabilities/src/plan_pick_place_capability.cpp]: How is this filled into or passed to the default grasp provider?

File: core/src/tasks/pick_place_task.cpp [hunk 15]
```diff
@@ -131,254 +131,54 @@ void PickPlaceTask::init(const Parameters& parameters)
...
    *                                                    *
    *****************************************************/
-  {
-    auto stage = std::make_unique<stages::MoveTo>("move home", sampling_planner);
-    stage->properties().configureInitFrom(Stage::PARENT, { "group" });
-    stage->setGoal(parameters.arm_home_pose_);
-    stage->restrictDirection(stages::MoveTo::FORWARD);
-    t.add(std::move(stage));
-  }
+  // {
```
Comment (felixvd) [inline core/src/tasks/pick_place_task.cpp]: Commented code bad, bad

File: msgs/action/PlanPickPlace.action [hunk 16]
```diff
@@ -1,7 +1,23 @@
...
+
+# object and surfaces
 string object_id
-string end_effector
+string[] support_surfaces
+
+# picking
+string grasp_provider_plugin_name
+moveit_msgs/Grasp grasp
+geometry_msgs/Pose grasp_frame_transform
```
Comment (felixvd) [inline msgs/action/PlanPickPlace.action]: What does this mean? Please add all the explanation to this `.action` file, it's the main piece of documentation for the user. Make sure it has all the details and doesn't require detective work.

File: msgs/action/PlanPickPlace.action [hunk 17]
```diff
@@ -1,7 +1,23 @@
...
+string eef_name
+string hand_frame
+
+# object and surfaces
 string object_id
-string end_effector
+string[] support_surfaces
+
+# picking
+string grasp_provider_plugin_name
```
Comment (felixvd) [inline msgs/action/PlanPickPlace.action]: Same for this. As an end user, what do I have to do about this? What is the default?

File: core/src/tasks/pick_place_task.cpp (hunk 13 above)
Comment (felixvd) [reply core/src/tasks/pick_place_task.cpp]: Oh, I think I might have misunderstood. This is actually used in the capability, right? Hmmm.

File: core/include/moveit/task_constructor/stages/pick.h [hunk 18]
```diff
@@ -32,19 +32,27 @@
...
 
 namespace moveit {
 namespace task_constructor {
 
 namespace solvers {
 MOVEIT_CLASS_FORWARD(CartesianPath)
+MOVEIT_CLASS_FORWARD(PipelinePlanner)
 }
 
 namespace stages {
```
Comment (felixvd) [inline core/include/moveit/task_constructor/stages/pick.h]: I would add that the `Pick` and `Place` stages inherit from this `PickPlaceBase` class, and that it's not meant to be used directly.

File: core/src/tasks/pick_place_task.cpp [hunk 19]
```diff
@@ -131,254 +131,54 @@ void PickPlaceTask::init(const Parameters& parameters)
...
-    stage->properties().configureInitFrom(Stage::PARENT);
+    auto stage = std::make_unique<stages::Pick>("Pick object", parameters.grasp_provider_plugin_name_);
+    stage->properties().property("eef_group").configureInitFrom(Stage::PARENT, "hand");
+    stage->properties().property("eef_parent_group").configureInitFrom(Stage::PARENT, "group");
+    stage->setObject(parameters.object_name_);
+    stage->setEndEffector(parameters.eef_name_);
+    stage->setEndEffectorOpenClose(parameters.hand_open_pose_, parameters.hand_close_pose_);
+    stage->setSupportSurfaces(parameters.support_surfaces_);
+    stage->setIKFrame(parameters.grasp_frame_transform_, parameters.hand_frame_);
+    stage->GraspProviderPlugin()->properties().set("angle_delta", M_PI / 12);  // Set plugin-specific properties
```
Comment (felixvd) [inline core/src/tasks/pick_place_task.cpp]: What about the `PlaceProviderPlugin`? And why is this angle hardcoded?

**Review by felixvd — COMMENTED** [review]
Good progress, but it seems like there is still some work to do. I am especially unsure about the large changes to `pick_place_task.cpp`, since that file was likely useful as an example usage of MTC, it seems less readable now, and it seems that some things are hard-coded. 

Other thoughts: 

1. I don't really understand the GraspProviderPlugin and PlaceProviderPlugin as a user. What remains to be implemented for that? What is the easiest way to "just use" the capability?
2. This capability exposes the combined Pick-And-Place task, but not Pick and Place separately. Those should also be ported to MTC though, so we can deprecate the current pipeline. Adding two flags `plan_pick_only` and `plan_place_only` would probably solve this. We can set them as needed in the `pick` and `place` functions of the `move_group_interface`.

Really looking forward to seeing this work though!

**Comment by karolyartur** [comment]
@felixvd 
Sorry, it took more time than I expected. I wanted to propagate the property for the [hand open and close poses](https://github.com/karolyartur/moveit_task_constructor/blob/ea872dc6687bc6412611aa5a7bed80f74996b2d9/core/src/stages/grasp_provider.cpp#L135) from here (inside the Pick serial container) to the [MoveTo stage](https://github.com/karolyartur/moveit_task_constructor/pull/1/commits/ea872dc6687bc6412611aa5a7bed80f74996b2d9#diff-c578d8ba9c50bc52592ac22e95a30bae816083d8551cac84b3780b35fa730288L141) at the beginning of the task, but for this I would have to get the properties from inside the serial container to propagate out. This I don't know how to do, so I just ended up deleting the external MoveTo stage, expecting, that if you want to do a pick, you need to open the gripper for yourself as it is not included in the Pick pipeline. (Although a check, if it is actually open, can be included in the applicability filter. But I think if it is not open it will fail when it collides with the object during grasping anyway.)

I didn't have time to implement the same thing for the Place part but after this is done, it will be significantly faster.

The [script](https://gist.github.com/karolyartur/7785b7e9399b32a9974be0cfa04ba789) with which I tested the functionality.
For this to work, start the o2ac demo scene and spawn a box collision object with sizes 0.05 x 0.05 x 0.05 using RViz. Drag it above the table and publish. After that if you call this script you should be able to see it planning the Pick.
